Bug 1872469: Port login tests from protractor to Cypress

dtaylor113 opened the pull request:
This PR removes most login tests from protractor and adds them to Cypress.

spadgett commented on frontend/packages/integration-tests-cypress/tests/app/auth-multi-user-login.spec.ts:
```diff
@@ -0,0 +1,63 @@
+
+  it(`log in as kubeadmin user, verify sidenav menus`, () => {
+    cy.login(KUBEADMIN_IDP, KUBEADMIN_USERNAME, Cypress.env('BRIDGE_KUBEADMIN_PASSWORD'));
+    cy.visit(``);
```
```suggestion
    cy.visit('');
```

dtaylor113 replied:
fixed

spadgett reviewed (changes requested):
We need to update `test-prow.sh` to run the Cypress login spec when `login` is passed as a parameter. See

https://github.com/openshift/release/blob/024c148b0c0358eb8464c6f0ad5e31b6c304c002/ci-operator/config/openshift/cluster-authentication-operator/openshift-cluster-authentication-operator-master.yaml#L74

spadgett commented:
Hey, Dave. I'm glad we're porting these to Cypress, but we also want to make sure there isn't a real bug in logout that is causing the Protractor tests to fail.

spadgett commented:
Maybe we'll be able to reproduce with the Cypress tests and get a video to help debug.

dtaylor113 commented:
> Maybe we'll be able to reproduce with the Cypress tests and get a video to help debug.

yes, this PR is just another option and debugging tool, which is why it's a WIP :-)
Wanted to do something to progress https://bugzilla.redhat.com/show_bug.cgi?id=1872469

dtaylor113 commented on frontend/packages/integration-tests-cypress/tests/app/auth-multi-user-login.spec.ts:
```diff
@@ -0,0 +1,63 @@
+import { checkErrors } from '../../support';
```
fixed

dtaylor113 commented:
/retest

dtaylor113 commented:
Just verified protractor's login tests as well as these new Cypress login tests worked multiple times when run locally.

dtaylor113 commented on frontend/packages/integration-tests-cypress/tests/app/auth-multi-user-login.spec.ts:
```diff
@@ -0,0 +1,63 @@
+      nav.sidenav.switcher.changePerspectiveTo('Administrator');
+      nav.sidenav.switcher.shouldHaveText('Administrator');
+
+      cy.log('does not show admin nav items in Administration to test user');
```
Fixed. Adjusted so that when running Cypress we expose Redux store for debugging ie: `window.store = store`, along with [cypress-pipe](https://github.com/NicholasBoll/cypress-pipe) plug in, which enables us to retry until all feature flags used in the side nav menus are defined, before attempting to test the side nav menu contents.

dtaylor113 commented:
> We need to update `test-prow.sh` to run the Cypress login spec when `login` is passed as a parameter.

Added to bottom of `test-prow-e2e.sh`:
```
if ["$1" == "login" ]; then
  ./test-cypress.sh 'tests/app/auth-multiuser-login.spec.ts'
fi
```
..and updated `test-cypress.sh` to run `--spec` for any passed in arg.

dtaylor113 commented:
/retest

dtaylor113 commented:
/retest

dtaylor113 commented:
/retest

spadgett commented on test-prow-e2e.sh:
```diff
@@ -38,3 +38,6 @@ CHROME_VERSION=$(google-chrome --version) ./test-gui.sh "${1:-e2e}"
 if [ $# -eq 0 ] || [ "$1" == "release" ]; then
   ./test-cypress.sh
 fi
+if [ $# -eq 1 ] && ["$1" == "login" ]; then
+  ./test-cypress.sh 'tests/app/auth-multiuser-login.spec.ts'
+fi
```
We can skip running Protractor entirely if `"$1" == "login"`

dtaylor113 replied:
ok, updated `test-prow-e2e.sh`

spadgett commented on frontend/packages/integration-tests-cypress/reporter-config.json:
```diff
@@ -9,6 +9,7 @@
     "html": false,
-    "json": true
+    "json": true,
+    "quiet" : true
```
What output is this suppressing? Generally we want verbose output in CI for troubleshooting.

dtaylor113 replied:
This removes:
`[mochawesome] Report JSON saved to /home/dtaylor/repos/openshift/origin/src/github.com/openshift/console/frontend/gui_test_screenshots/cypress_report_004.json`

dtaylor113 replied:
Ok,  pulled unrelated changes.

spadgett commented on frontend/packages/integration-tests-cypress/plugins/index.js:
```diff
@@ -48,6 +48,12 @@ module.exports = (on, config) => {
+  config.env.BRIDGE_HTPASSWD_USERNAME = process.env.BRIDGE_HTPASSWD_USERNAME;
+  config.env.BRIDGE_HTPASSWD_PASSWORD = process.env.BRIDGE_HTPASSWD_PASSWORD;
   config.env.BRIDGE_KUBEADMIN_PASSWORD = process.env.BRIDGE_KUBEADMIN_PASSWORD;
+  config.env.appHost = `${process.env.BRIDGE_BASE_ADDRESS || 'http://localhost:9000'}${(
```
This looks like it's the same as `baseUrl`. Can we have one config value tracking?

dtaylor113 replied:
wow, talk about cut & paste blinders :-)  sorry, removed `appHost`.

spadgett commented on frontend/packages/integration-tests-cypress/support/index.ts:
```diff
@@ -20,6 +20,11 @@ Cypress.Cookies.defaults({
   preserve: ['openshift-session-token', 'csrf-token'],
 });
 
+Cypress.Commands.overwrite('log', (originalFn, message) => {
```
I'm undecided on this change since it's hard to tell what's a test and what's not looking at the console output. For debugging, we have the screenshot and video where you can see the log statements.

dtaylor113 replied:
The test output has a checkmark in front of it, log statements do not.  IMO this really helps see what is happening in headless mode.

dtaylor113 replied:
Ok, pulled unrelated changes.

spadgett reviewed (commented):
Thanks @dtaylor113. Let's pull the unrelated changes into a separate PR. I don't want to block an important test fix from code comments on the other changes.

dtaylor113 commented on frontend/packages/integration-tests-cypress/cypress.json:
```diff
@@ -12,7 +12,8 @@
+    "runMode": 1,
+    "openMode": 0
+  },
+  "videoCompression": 0
```
I'm not sure how much larger, just noticed video playback was choppy and was missing transitions.

dtaylor113 replied:
Ok, pulled unrelated changes.

spadgett commented on test-prow-e2e.sh:
```diff
@@ -32,9 +32,14 @@ oc patch oauths cluster --patch "$(cat ./frontend/integration-tests/data/patch-h
 export DBUS_SESSION_BUS_ADDRESS
 
-CHROME_VERSION=$(google-chrome --version) ./test-gui.sh "${1:-e2e}"
+if [ $# -eq 0 ] || [ "$1" != "login" ]; then
+  CHROME_VERSION=$(google-chrome --version) ./test-gui.sh "${1:-e2e}"
+fi
```
```suggestion
SCENARIO="${1:=e2e}"
# Only run Cypress tests for the login scenario.
if [ "$SCENARIO" != "login" ]; then
  CHROME_VERSION=$(google-chrome --version) ./test-gui.sh "$SCENARIO"
fi
```

Then you can use `$SCENARIO` again below instead of `[ $# -eq 0 ] || [ "$1" == "release" ]`, etc.

dtaylor113 replied:
thanks, fixed

dtaylor113 commented on test-cypress.sh:
```diff
@@ -8,4 +8,8 @@ function generateReport {
 trap generateReport EXIT
 
-yarn run test-cypress-headless
+if [ $# -gt 0 ] && [ -n "$1" ]; then
```
Unfortunately, I get `$1: unbound variable` when I don't provide a param.

spadgett replied:
@dtaylor113 Did you use `"{$1:-}"` or just `"$1"`?

dtaylor113 replied:
`{$1:-}`

dtaylor113 commented on frontend/packages/integration-tests-cypress/tests/app/auth-multiuser-login.spec.ts:
```diff
@@ -0,0 +1,63 @@
+      nav.sidenav.switcher.changePerspectiveTo('Administrator');
+      nav.sidenav.switcher.shouldHaveText('Administrator');
+
+      cy.log('does not show admin nav items in Administration to test user');
```
ok, fixed

spadgett commented on test-prow-e2e.sh:
```diff
@@ -32,9 +32,16 @@ oc patch oauths cluster --patch "$(cat ./frontend/integration-tests/data/patch-h
+elif [ "$SCENARIO" == "login" ]; then
+  ./test-cypress.sh 'tests/app/auth-multiuser-login.spec.ts'
+else
+  ./test-cypress.sh $SCENARIO
```
This looks wrong... We should not run any cypress tests for other scenarios. You can remove this last `else`.

dtaylor113 replied:
Hmm..I can change, but protractor will run with an unknown scenario.

dtaylor113 replied:
ok, updated

spadgett commented on test-prow-e2e.sh:
```diff
@@ -32,9 +32,16 @@ oc patch oauths cluster --patch "$(cat ./frontend/integration-tests/data/patch-h
+elif [ "$SCENARIO" == "login" ]; then
+  ./test-cypress.sh 'tests/app/auth-multiuser-login.spec.ts'
+else
+  echo "Cypress: unknown parameter: $1"
```
Don't echo here. This is expected if you're running a protractor suite. Just remove the else.

dtaylor113 replied:
Ok, removed

spadgett reviewed (approved):
/lgtm

Thanks @dtaylor113 !

dtaylor113 commented:
/retest
